enterprise/server/backends/pebble_cache/pebble_cache.go

```diff
@@ -2435,12 +2453,120 @@ func newPartitionEvictor(part disk.Partition, fileStorer filestore.Store, blobDi
 	return pe, nil
 }
 
+func (e *partitionEvictor) startSampleGenerator(quitChan chan struct{}) {
+	eg := &errgroup.Group{}
+	eg.Go(func() error {
+		return e.generateSamplesForEviction(quitChan)
+	})
+	eg.Wait()
+	// Drain samples chan before exiting
+	for len(e.samples) > 0 {
+		<-e.samples
+	}
+	close(e.samples)
+}
+
+func (e *partitionEvictor) generateSamplesForEviction(quitChan chan struct{}) error {
+	db, err := e.dbGetter.DB()
+	if err != nil {
+		log.Warningf("cannot generate samples for eviction: failed to get db: %s", err)
+		return err
+	}
+	defer db.Close()
+	start, end := keyRange([]byte(e.partitionKeyPrefix() + "/"))
+	iter := db.NewIter(&pebble.IterOptions{
+		LowerBound: start,
+		UpperBound: end,
+	})
+	// We update the iter variable later on, so we need to wrap the Close call
+	// in a func to operate on the correct iterator instance.
+	defer func() {
+		iter.Close()
+	}()
+
+	totalCount := 0
+
+	// Files are kept in random order (because they are keyed by digest), so
+	// instead of doing a new seek for every random sample we will seek once
+	// and just read forward, yielding digests until we've found enough.
+
+	for {
+		select {
+		case <-quitChan:
+			return nil
+		default:
+		}
+		// Refersh the iterator once a while
+		if totalCount > e.samplesPerBatch {
+			totalCount = 0
+			newIter := db.NewIter(&pebble.IterOptions{
+				LowerBound: start,
+				UpperBound: end,
+			})
+			iter.Close()
+			iter = newIter
+		}
+		totalCount += 1
+		if !iter.Valid() {
+			// This should happen once every totalCount times or when
+			// we exausted the iter.
+			randomKey, err := e.randomKey(64)
+			if err != nil {
+				log.Warningf("cannot generate samples for eviction: failed to get random key: %s", err)
+				return err
+			}
+			valid := iter.SeekGE(randomKey)
+			if !valid {
+				continue
+			}
+		}
+		var key filestore.PebbleKey
+		if _, err := key.FromBytes(iter.Key()); err != nil {
+			log.Warningf("cannot generate sample for eviction, skipping: failed to read key: %s", err)
+			continue
+		}
+
+		fileMetadata := &rfpb.FileMetadata{}
+		unlockFn := e.locker.RLock(key.LockID())
```

Comment:
this is copied over from the previous code but as far as i can tell this locking does nothing whatsoever?